src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -25,37 +28,173 @@ object PV63 {
   }
 
-  case class GetNodeData(hashes: Seq[ByteString]) extends Message {
+  case class GetNodeData(commonMptHashes: Seq[ByteString]) extends Message {
     override def code: Int = GetNodeData.code
 
     override def toString: String = {
       s"""GetNodeData{
-         |hashes: ${hashes.map(e => Hex.toHexString(e.toArray[Byte]))}
+         |hashes: ${commonMptHashes.map(e => Hex.toHexString(e.toArray[Byte]))}
+         |}
+       """.stripMargin
+    }
+  }
+
+  object Account {
+    implicit val rlpEndDec = new RLPEncoder[Account] with RLPDecoder[Account] {
+      override def encode(obj: Account): RLPEncodeable = {
+        import obj._
+        RLPList(nonce, balance, byteStringEncDec.encode(storageRoot), byteStringEncDec.encode(codeHash))
+      }
+
+      override def decode(rlp: RLPEncodeable): Account = rlp match {
+        case RLPList(nonce, balance, storageRoot, codeHash) =>
+          Account(nonce, balance, byteStringEncDec.decode(storageRoot), byteStringEncDec.decode(codeHash))
+        case _ => throw new RuntimeException("Cannot decode Account")
+      }
+    }
+  }
+
+  case class Account(nonce: BigInt, balance: BigInt, storageRoot: ByteString, codeHash: ByteString) {
+    override def toString: String = {
+      s"""Account{
+         |nonce: $nonce
+         |balance: $balance wei
+         |storageRoot: ${Hex.toHexString(storageRoot.toArray[Byte])}
+         |codeHash: ${Hex.toHexString(codeHash.toArray[Byte])}
          |}
        """.stripMargin
     }
   }
 
+  object MptNode {
+    implicit val rlpEndDec = new RLPEncoder[MptNode] with RLPDecoder[MptNode] {
+      override def encode(obj: MptNode): RLPEncodeable = {
+        obj match {
+          case n: MptLeaf =>
+            import n._
+            RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = true)), value.fold(a => rlpEncode[Account](a), b => b: RLPEncodeable))
+          case n: MptExtension =>
+            import n._
+            RLPList(RLPValue(hpEncode(keyNibbles.toArray[Byte], isLeaf = false)), child.fold(_.hash, _.value): RLPEncodeable)
+          case n: MptBranch =>
+            import n._
+            RLPList(children.map(e => e.fold(_.hash, _.value)).map(e => byteStringEncDec.encode(e)) :+ (terminator: RLPEncodeable): _*)
+        }
+      }
+
+      override def decode(rlp: RLPEncodeable): MptNode = rlp match {
+        case rlpList: RLPList if rlpList.items.length == 17 =>
+          MptBranch(rlpList.items.take(16).map { bytes =>
+            val v = rlpDecode[ByteString](bytes)
+            if (v.nonEmpty && v.length < 32) {
+              Right(MptValue(v))
+            } else {
+              Left(MptHash(v))
+            }
+          }, byteStringEncDec.decode(rlpList.items(16)))
+        case RLPList(hpEncoded, value) =>
+          hpDecode(hpEncoded: Array[Byte]) match {
+            case (decoded, true) =>
+              MptLeaf(ByteString(decoded),
+                Try(Left(rlpDecode[Account](value: Array[Byte])))
```

Comment:
How about keeping this message more low-level, i.e according to documentation https://github.com/ethereum/wiki/wiki/Ethereum-Wire-Protocol) it returns a list of Bytes (instead of Either[MptNode, ByteString]) - I think it's what Alan suggested as RawNodeResponse. Then we might have one place (actor) that requests NodeData and receives such responses - it can store a map of what was requested and decode accordingly without the need to modify the Message decode interface.